[csharp] Grpc.Tools documentation for sharing proto files between projects and nuget packages (#34521)
grpc/grpc: src/csharp/BUILD-INTEGRATION.md

For more examples see the example project files in GitHub: https://github.com/grpc/grpc-dotnet/tree/master/examples

## Sharing `.proto` files between multiple projects (in the same VS solution)

> If you have one or more `.proto` files that you want to use in multiple project then
> you may create a shared class library that compiled the generated the code.
> The other projects in the solution then can reference this shared class library instead of each project having to compile
> the same `.proto` files.

Reviewer: ````suggestion
It's common to want to share `.proto` files between projects. For example, a gRPC client
and gRPC server share the same contract. It is preferable to share contracts without
creating copies because copies can go out of sync over time.

There are a couple of ways to use `.proto` files in multiple projects with copies and duplication:

* Sharing `.proto` files between projects with MSBuild links.
* Generating code in a class library and sharing the library.

### Sharing `.proto` with MSBuild links

`.proto` files can be placed in a shared location and referenced by multiple projects
using MSBuild's [`Link` or `LinkBase` settings](https://learn.microsoft.com/visualstudio/msbuild/common-msbuild-item-metadata).

```xml
<ItemGroup>
   <Protobuf Include="..\Protos\greet.proto" GrpcServices="None" Link="Protos\greet.proto"/>
</ItemGroup>
```

In the example above, `greet.proto` is located in a shared `Protos` directory outside of
the project directory. Multiple projects can reference the proto file.

### Generating code in a class library

Create a class library that references `.proto` files and contains generated code. The other
projects in the solution can then reference this shared class library instead of each project
having to compile the same `.proto` files.
````

---

## <a name="proto-only-nuget"></a>Including `.proto` files in NuGet packages

There might be occassions when you are given a NuGet package that contains
`.proto` files that you wish to use in your own project, or you may wish to 
package your own `.proto` files in a NuGet package for others to use.

There is no automatic way for `Grpc.Tools` to locate and include `.proto`
files from other NuGet packages. Below is a suggested convention to use
when creating NuGet packages that contain `.proto` files.

__Note:__ This is not the same as a NuGet package providing a library built from

> the code generated from the `.proto` files. Below just describes how to provide
> the uncompiled `.proto` files in a NuGet package.

Reviewer: ```suggestion
the code generated from the `.proto` files. Below just describes how to provide
the uncompiled `.proto` files in a NuGet package and have `Grpc.Tools` automatically
use them.
```

### Creating the NuGet package

The NuGet package should:
- provide the `.proto` files in a `content\protos` subdirectory in the package
- provide a `packagename.targets` file in the `build` subdirectory in the package that:
  - defines an MSBuild property giving the path to the `.proto` files in the
    installed package
  - conditionally updates the `Protobuf_StandardImportsPath` property with the
    above path so that the files can be found by the protocol buffers compiler
    - it should be made optional forcing users to *opt in* to including
      the `.proto` files

For example, for the package `My.Example.Protos`:

My.Example.Protos.nuspec:

> ```XML

Reviewer: ```suggestion
```xml
```

<?xml version="1.0"?>
<package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
  <metadata>
    <id>My.Example.Protos</id>
    <version>1.0.0</version>
    <title>Example package containing proto files</title>
    <authors>author</authors>
    <owners>owner</owners>
    <licenseUrl>license url</licenseUrl>
    <projectUrl>project url</projectUrl>
    <description>See project site for more info.</description>
    <summary>Example package containing proto files.</summary>
    <releaseNotes>Example package containing proto files</releaseNotes>
    <copyright>Copyright 2023, My Company.</copyright>
  </metadata>
  
  <files>
    <!-- copy the My.Example.Protos.targets file for MSBuild integration -->
    <file src="build\**" target="build" />
    <!-- copy the .proto files into the package -->
    <file src="proto\**" target="content\protos" />
  </files>
</package>
```

My.Example.Protos.targets:

> ```XML

Reviewer: ```suggestion
```xml
```

  </PropertyGroup>

  <!-- Run immediately before the Protobuf_BeforeCompile extension point. -->
  <!-- Only include protos if project has set <IncludeMyExampleProtosProtos> 
       property to true. -->
  <Target Name="MyExampleProtos_BeforeCompile"
          BeforeTargets="Protobuf_BeforeCompile"
          Condition=" '$(IncludeMyExampleProtosProtos)' == 'true' ">
    <PropertyGroup>
      <!-- Add proto files by including path in Protobuf_StandardImportsPath.
           This path is passed to protoc via the -I option -->
      <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);$(MyPackage_ProtosPath)</Protobuf_StandardImportsPath>
    </PropertyGroup>

    <!-- These message are not required but included here for diagnostics -->
    <Message Text="Included proto files at $(MyExampleProtos_ProtosPath) in import path." Importance="high" />
    <Message Text="Updated proto imports path: $(Protobuf_StandardImportsPath)" Importance="high" />
  </Target>
</Project>
```

### Using the NuGet package

The project needs to add the package containing the `.proto` files:

> ```XML

Reviewer: ```suggestion
```xml
```

<PackageReference Include="My.Example.Protos" Version="1.0.0" />
```

If the project only wants to compile the `.proto` files included in the package
then all it needs to do is add the `<Protobuf>` items using the property defined
in the package for the path to the files. For example, if the NuGet package contained
the file `greet.proto`, then the project should add:

> ```XML

Reviewer: ```suggestion
```xml
```

<Protobuf Include="$(MyExampleProtos_ProtosPath)/greet.proto" />
```

However if the provided `.proto` files are to be *imported* by the projects own `.proto`
files then the `Protobuf_StandardImportsPath` needs updated to add the directory
containing the package's files.  This is done by setting to `true` the property
used in the package. For example, if the project has the local `.proto` file
`my_services.proto` and it imported a file from the package `common_message.proto`,
then:

> ```XML

Reviewer: ```suggestion
```xml
```
